kernel/kernel-api/src/main/java/io/delta/kernel/internal/replay/LogReplay.java

```diff
@@ -296,4 +310,50 @@ private Optional<Long> loadLatestTransactionVersion(Engine engine, String applic
 
     return Optional.empty();
   }
+
+  /**
+   * Retrieves a map of domainName to {@link DomainMetadata} from the log files.
+   *
+   * <p>Now loading domain metadata requires an additional round of log replay so this is done
+   * lazily only when domain metadata is requested. We might want to merge this into {@link
+   * #loadTableProtocolAndMetadata}.
+   *
+   * @param engine The engine used to process the log files.
+   * @return A map where the keys are domain names and the values are the corresponding {@link
+   *     DomainMetadata} objects.
+   * @throws RuntimeException if an I/O error occurs while closing the iterator.
+   */
+  public Map<String, DomainMetadata> loadDomainMetadataMap(Engine engine) {
+
+    Map<String, DomainMetadata> domainMetadataMap = new HashMap<>();
+
+    try (CloseableIterator<ActionWrapper> reverseIter =
+        new ActionsIterator(
+            engine,
+            logSegment.allLogFilesReversed(),
+            DOMAIN_METADATA_READ_SCHEMA,
+            Optional.empty())) {
+      while (reverseIter.hasNext()) {
+
+        final ColumnarBatch columnarBatch = reverseIter.next().getColumnarBatch();
+        assert (columnarBatch.getSchema().equals(DOMAIN_METADATA_READ_SCHEMA));
+
+        final ColumnVector dmVector = columnarBatch.getColumnVector(0);
+        for (int rowId = 0; rowId < dmVector.getSize(); rowId++) {
+          if (!dmVector.isNullAt(rowId)) {
+            DomainMetadata dm = DomainMetadata.fromColumnVector(dmVector, rowId);
+            // We keep the latest version of each domain.
+            if (dm != null && !domainMetadataMap.containsKey(dm.getDomain())) {
+              domainMetadataMap.put(dm.getDomain(), dm);
+            }
+          }
+        }
+      }
+
+    } catch (IOException ex) {
+      throw new RuntimeException("Could not close iterator", ex);
+    }
+
+    return domainMetadataMap;
```

Comment:
declare the map inside the try please, and return it inside the try

Reply:
Updated.